File: src/tdg/tdg.cpp
#include "tdg/tdg.h"

#include <spdlog/spdlog.h>

#include <algorithm>
#include <fstream>
#include <sstream>
#include <stdexcept>

#include "../json/json.h"

namespace tdg {

namespace {

// Logs per-core task priority ordering for debugging and diagnostics.
std::string format_core_priority_order(
    int core_id, const std::vector<std::string>& tasks,
    const std::unordered_map<std::string, int>& tasks_priority,
    const std::string& prefix) {
  std::ostringstream oss;
  oss << prefix << " Core " << core_id << " priority order: ";

  bool first = true;
  for (const auto& task : tasks) {
    if (!first) {
      oss << " > ";
    }
    first = false;
    oss << task << "(" << tasks_priority.at(task) << ")";
  }

  if (first) {
    oss << "(none)";
  }

  return oss.str();
}

// Registers a parsed node into the TDG lookup tables.
void register_node(TDG& tdg, const NodeType& node_type, bool log_node) {
  visit_node(node_type, [&](const auto& node) {
    using Node = std::decay_t<decltype(node)>;

    if constexpr (std::is_same_v<Node, TaskNode>) {
      tdg.all_task.emplace_back(node_type);
      tdg.tasks_priority.emplace(node.name, node.priority);
      tdg.nodes_type.emplace(node.name, node_type);
      tdg.vertexes_type.emplace(node.name, TDGVertexType::TASK);
      tdg.tasks_type.emplace(node.name, TaskType::NORMAL);

      for (const auto& lock : node.lock) {
        tdg.lock_set.insert(lock);
        tdg.task_locks_map[node.name].push_back(lock);
      }

      if (log_node) {
        spdlog::info("[TDG] Node '{}' -> task (priority={}, core={})",
                     node.name, node.priority, node.core);
      }
      return;
    }

    if constexpr (std::is_same_v<Node, ForkTask>) {
      tdg.nodes_type.emplace(node.name, node_type);
      tdg.vertexes_type.emplace(node.name, TDGVertexType::FORK);
      if (log_node) {
        spdlog::info("[TDG] Node '{}' -> fork", node.name);
      }
      return;
    }

    if constexpr (std::is_same_v<Node, JoinTask>) {
      tdg.nodes_type.emplace(node.name, node_type);
      tdg.vertexes_type.emplace(node.name, TDGVertexType::JOIN);
      if (log_node) {
        spdlog::info("[TDG] Node '{}' -> join", node.name);
      }
      return;
    }

    if constexpr (std::is_same_v<Node, EmptyTask>) {
      tdg.nodes_type.emplace(node.name, node_type);
      tdg.vertexes_type.emplace(node.name, TDGVertexType::EMPTY);
      if (log_node) {
        spdlog::info("[TDG] Node '{}' -> empty", node.name);
      }
    }
  });
}

// Populates TDG fields from a successfully parsed JSON graph.
void load_from_parser(TDG& tdg, const parse::Parser& parser, bool log_nodes) {
  tdg.num_cpus = parser.get_num_cpus();
  tdg.cores_per_cpu = parser.get_cores_per_cpu();
  tdg.policy = parser.get_policy();
  tdg.start_tasks = parser.get_start_tasks();
  tdg.end_tasks = parser.get_end_tasks();
  tdg.periodic_tasks = parser.get_periodic_tasks();

  for (const auto& json_node : parser.get_nodes()) {
    register_node(tdg, json_node.to_node_type(), log_nodes);
  }

  tdg.tdg_edges.clear();
  tdg.tdg_edges.reserve(parser.get_edges().size());
  for (const auto& edge : parser.get_edges()) {
    tdg.tdg_edges.push_back({edge.source, edge.target, edge.label, edge.style});
    spdlog::debug("[TDG] Edge: {} -> {} (style={})", edge.source, edge.target,
                  edge.style);
  }
}

}  // namespace

void TDG::parse_json(const std::string& json_file) {
  spdlog::info("[TDG] Starting JSON parsing: {}", json_file);

  parse::Parser parser;
  const auto result = parser.parse_file(json_file);
  if (!result.success) {
    spdlog::error("[TDG] JSON parsing failed: {}", result.error_message);
    return;
  }

  spdlog::info("[TDG] Configuration: {} CPUs, {} cores per CPU",
               parser.get_num_cpus(), parser.get_cores_per_cpu());

  load_from_parser(*this, parser, /*log_nodes=*/true);

  spdlog::info("[TDG] JSON parsing completed: {} nodes, {} edges",
               nodes_type.size(), tdg_edges.size());
}

void TDG::parse_json_string(const std::string& json_content) {
  parse::Parser parser;
  const auto result = parser.parse_string(json_content);
  if (!result.success) {
    throw std::runtime_error("JSON parsing failed: " + result.error_message);
  }

  load_from_parser(*this, parser, /*log_nodes=*/false);
}

std::string TDG::to_dot_string() const {
  std::ostringstream oss;

  oss << "digraph G {\n";

  for (const auto& [name, node] : nodes_type) {
    oss << "    " << name << " [label = \"" << parse::node_to_dot_label(node)
        << "\";];\n";
  }

  for (const auto& edge : tdg_edges) {
    oss << "    " << edge.source << " -> " << edge.target;
    if (!edge.label.empty() || !edge.style.empty()) {
      oss << " [";
      if (!edge.label.empty()) {
        oss << "xlabel = \"" << edge.label << "\"";
      }
      if (!edge.style.empty()) {
        if (!edge.label.empty()) {
          oss << "; ";
        }
        oss << "style = \"" << edge.style << "\"";
      }
      oss << ";]";
    }
    oss << ";\n";
  }

  oss << "}\n";
  return oss.str();
}

void TDG::export_to_dot(const std::string& output_path) {
  spdlog::info("[DOT] Exporting to: {}", output_path);

  std::ofstream file(output_path);
  if (!file.is_open()) {
    spdlog::error("[DOT] Failed to create file: {}", output_path);
    return;
  }

  file << to_dot_string();

  spdlog::info("[DOT] Exported {} nodes, {} edges to {}", nodes_type.size(),
               tdg_edges.size(), output_path);
}

std::unordered_map<int, std::vector<std::string>> TDG::classify_priority() {
  std::unordered_map<int, std::vector<std::string>> core_task;

  for (const auto& task : all_task) {
    if (const auto* task_node = as_task_node(task)) {
      tasks_config.emplace(task_node->name,
                           TaskConfig{task_node->core, task_node->priority,
                                      task_node->time, task_node->lock});
      core_task[task_node->core].push_back(task_node->name);
    }
  }

  for (auto& [core_id, tasks] : core_task) {
    std::sort(tasks.begin(), tasks.end(),
              [&](const std::string& left, const std::string& right) {
                return tasks_priority.at(left) > tasks_priority.at(right);
              });
  }

  for (const auto& [core_id, tasks] : core_task) {
    spdlog::info("{}", format_core_priority_order(core_id, tasks,
                                                  tasks_priority, "[TDG]"));
  }

  return core_task;
}

}  // namespace tdg

File: rust/src/tdg.rs
//! Task Dependency Graph model (port of `src/tdg/tdg.cpp`).

use crate::json::Parser;
use crate::types::{
    NodeType, PeriodicBinding, SchedulePolicy, StartBinding, TdgEdge, TdgVertexType, TaskConfig,
    TaskType,
};
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Default)]
pub struct TDG {
    pub num_cpus: i32,
    pub cores_per_cpu: i32,
    pub task_place_capacity: i32,
    pub policy: SchedulePolicy,
    /// All task nodes in insertion order.
    pub all_task: Vec<NodeType>,
    pub start_tasks: Vec<StartBinding>,
    pub end_tasks: Vec<String>,
    pub periodic_tasks: Vec<PeriodicBinding>,
    pub tasks_priority: HashMap<String, i32>,
    pub vertexes_type: HashMap<String, TdgVertexType>,
    pub nodes_type: HashMap<String, NodeType>,
    pub tasks_type: HashMap<String, TaskType>,
    pub lock_set: HashSet<String>,
    pub task_locks_map: HashMap<String, Vec<String>>,
    pub tasks_config: HashMap<String, TaskConfig>,
    pub tdg_edges: Vec<TdgEdge>,
}

impl TDG {
    pub fn new(num_cpus: i32, cores_per_cpu: i32) -> Self {
        TDG {
            num_cpus,
            cores_per_cpu,
            task_place_capacity: 1,
            policy: SchedulePolicy::Fixed,
            ..Default::default()
        }
    }

    pub fn load_from_parser(&mut self, parser: &Parser, log_nodes: bool) {
        self.num_cpus = parser.get_num_cpus();
        self.cores_per_cpu = parser.get_cores_per_cpu();
        self.task_place_capacity = parser.get_task_place_capacity();
        self.policy = parser.get_policy();
        self.start_tasks = parser.get_start_tasks().to_vec();
        self.end_tasks = parser.get_end_tasks().to_vec();
        self.periodic_tasks = parser.get_periodic_tasks().to_vec();

        for json_node in parser.get_nodes() {
            self.register_node(&json_node.to_node_type(), log_nodes);
        }

        self.tdg_edges.clear();
        for edge in parser.get_edges() {
            self.tdg_edges.push(TdgEdge {
                source: edge.source.clone(),
                target: edge.target.clone(),
                label: edge.label.clone(),
                style: edge.style.clone(),
            });
        }
    }

    pub fn register_node(&mut self, node: &NodeType, log_node: bool) {
        match node {
            NodeType::Task(t) => {
                self.all_task.push(node.clone());
                self.tasks_priority.insert(t.name.clone(), t.priority);
                self.nodes_type.insert(t.name.clone(), node.clone());
                self.vertexes_type.insert(t.name.clone(), TdgVertexType::Task);
                self.tasks_type.insert(t.name.clone(), TaskType::Normal);

                for lock in &t.lock {
                    self.lock_set.insert(lock.clone());
                    self.task_locks_map
                        .entry(t.name.clone())
                        .or_default()
                        .push(lock.clone());
                }

                if log_node {
                    // no-op in library; CLI logs
                }
            }
            NodeType::Fork(f) => {
                self.nodes_type.insert(f.name.clone(), node.clone());
                self.vertexes_type.insert(f.name.clone(), TdgVertexType::Fork);
            }
            NodeType::Join(j) => {
                self.nodes_type.insert(j.name.clone(), node.clone());
                self.vertexes_type.insert(j.name.clone(), TdgVertexType::Join);
            }
            NodeType::Empty(e) => {
                self.nodes_type.insert(e.name.clone(), node.clone());
                self.vertexes_type
                    .insert(e.name.clone(), TdgVertexType::Empty);
            }
        }
    }

    /// Parses from a JSON file path (ignoring errors; CLI validates separately).
    pub fn parse_json(&mut self, json_file: &str) {
        let mut parser = Parser::new();
        if !parser.parse_file(json_file).success {
            return;
        }
        self.load_from_parser(&parser, true);
    }

    pub fn parse_json_string(&mut self, json_content: &str) -> Result<(), String> {
        let mut parser = Parser::new();
        let result = parser.parse_string(json_content);
        if !result.success {
            return Err(format!("JSON parsing failed: {}", result.error_message));
        }
        self.load_from_parser(&parser, false);
        Ok(())
    }

    pub fn to_dot_string(&self) -> String {
        let mut out = String::new();
        out.push_str("digraph G {\n");

        // Deterministic ordering (sorted by name) to keep output stable.
        let mut names: Vec<&String> = self.nodes_type.keys().collect();
        names.sort();
        for name in names {
            let node = &self.nodes_type[name];
            out.push_str(&format!(
                "    {} [label = \"{}\";];\n",
                name,
                crate::json::node_to_dot_label(node)
            ));
        }

        for edge in &self.tdg_edges {
            out.push_str(&format!("    {} -> {}", edge.source, edge.target));
            if !edge.label.is_empty() || !edge.style.is_empty() {
                out.push_str(" [");
                if !edge.label.is_empty() {
                    out.push_str(&format!("xlabel = \"{}\"", edge.label));
                }
                if !edge.style.is_empty() {
                    if !edge.label.is_empty() {
                        out.push_str("; ");
                    }
                    out.push_str(&format!("style = \"{}\"", edge.style));
                }
                out.push_str(";]");
            }
            out.push_str(";\n");
        }

        out.push_str("}\n");
        out
    }

    pub fn export_to_dot(&self, output_path: &str) -> bool {
        match std::fs::write(output_path, self.to_dot_string()) {
            Ok(_) => true,
            Err(_) => false,
        }
    }

    /// Core -> tasks sorted by descending priority.
    pub fn classify_priority(&mut self) -> HashMap<i32, Vec<String>> {
        let mut core_task: HashMap<i32, Vec<String>> = HashMap::new();

        for task in &self.all_task {
            if let Some(task_node) = task.as_task() {
                self.tasks_config.insert(
                    task_node.name.clone(),
                    TaskConfig {
                        core: task_node.core,
                        priority: task_node.priority,
                        times: task_node.time.clone(),
                        locks: task_node.lock.clone(),
                    },
                );
                core_task
                    .entry(task_node.core)
                    .or_default()
                    .push(task_node.name.clone());
            }
        }

        for tasks in core_task.values_mut() {
            tasks.sort_by(|left, right| {
                let pl = self.tasks_priority.get(left).copied().unwrap_or(0);
                let pr = self.tasks_priority.get(right).copied().unwrap_or(0);
                pr.cmp(&pl)
            });
        }

        core_task
    }
}
